tst/ProducerContinuousRetryTest.cpp

```diff
@@ -154,8 +154,179 @@ TEST_F(ProducerContinuousRetryTest, test_stream_recover_after_reset_connnection_
     mStreams[0] = INVALID_STREAM_HANDLE_VALUE;
 }
 
+TEST_F(ProducerContinuousRetryTest, recover_on_retriable_producer_error) {
+    STREAM_HANDLE streamHandle = INVALID_STREAM_HANDLE_VALUE;
+    UINT32 i;
+    UINT32 totalFragments = 10;
+    UINT32 totalFrames = totalFragments * TEST_FPS;
+    PRotatingStaticAuthCallbacks pAuth;
+
+    // block off any acks
+    mCurlWriteCallbackPassThrough = TRUE;
+    mWriteStatus = STATUS_NOT_IMPLEMENTED;
+
+    createDefaultProducerClient(FALSE, TEST_CREATE_STREAM_TIMEOUT, TRUE);
+
+    // Induce a new session on connection staleness
+    mStreamInfo.streamCaps.connectionStalenessDuration = 5 * HUNDREDS_OF_NANOS_IN_A_SECOND;
+
+    EXPECT_EQ(STATUS_SUCCESS, createTestStream(0, STREAMING_TYPE_REALTIME, TEST_MAX_STREAM_LATENCY, TEST_STREAM_BUFFER_DURATION));
+    streamHandle = mStreams[0];
+    EXPECT_TRUE(streamHandle != INVALID_STREAM_HANDLE_VALUE);
+
+    // Inject a fault
+    pAuth = (PRotatingStaticAuthCallbacks) mAuthCallbacks;
+    pAuth->retStatus = STATUS_INVALID_DESCRIBE_STREAM_RETURN_JSON;
+    pAuth->failCount = 0;
+    pAuth->recoverCount = 3; // 1 for main token, 1 for the security token for the first session
+    // and only then should fail for the last token.
+
+    for(i = 0; i < totalFrames; ++i) {
+        EXPECT_EQ(STATUS_SUCCESS, putKinesisVideoFrame(streamHandle, &mFrame));
+        updateFrame();
+        THREAD_SLEEP(mFrame.duration / 2); // speed up test
+    }
+
+    DLOGD("Stopping the stream with stream handle %" PRIu64, (UINT64) streamHandle);
+    EXPECT_EQ(STATUS_SUCCESS, stopKinesisVideoStream(streamHandle));
+    EXPECT_EQ(STATUS_SUCCESS, freeKinesisVideoStream(&streamHandle));
+    EXPECT_EQ(1, mConnectionStaleFnCount);
+    EXPECT_LT(1, mDescribeStreamFnCount);
```

Comment:
Tests failing because sometimes mDescribeStreamFnCount is 1.  Did you mean EXPECT_LE (which is less than or equal)?

Reply:
Yeah, it’s hard to get the timing precise as these are independent threads doing the network and recovery. Will change to le